File: samples/configdirector-android/java/src/main/java/com/configdirector/sample/java/SampleApplication.java
package com.configdirector.sample.java;

import android.app.Application;
import com.configdirector.ClientOptions;
import com.configdirector.ConfigDirectorClient;
import com.configdirector.ConnectionMode;
import com.configdirector.ConnectionOptions;
import com.configdirector.Metadata;

/**
 * The client belongs to the application: one instance, created once, initialized during startup and
 * shared by every screen.
 *
 * <p>Nothing closes it. It lives as long as the process does, and Android reclaims everything when
 * the process ends — {@code onTerminate} only runs on an emulator. Call {@code close()} when an app
 * wants the client gone before that, on sign-out for instance. The Close client button on the
 * screen does exactly that.
 */
public final class SampleApplication extends Application {

  // The SDK talks to a stubbed transport for now, which serves the sample's configs whatever the
  // key is. It becomes a real key from the ConfigDirector dashboard once the transports land.
  private static final String SAMPLE_SDK_KEY = "sample-client-sdk-key";

  private final SampleLog log = new SampleLog();

  private ConfigDirectorClient client;

  @Override
  public void onCreate() {
    super.onCreate();

    client =
        new ConfigDirectorClient(
            SAMPLE_SDK_KEY,
            ClientOptions.builder()
                .metadata(new Metadata("ConfigDirector Java Sample", "1.0"))
                .connection(
                    ConnectionOptions.builder()
                        .mode(ConnectionMode.STREAMING)
                        .timeoutMillis(3_000L)
                        .pausesWhileBackgrounded(true)
                        .build())
                .logger(new SampleLogger(log))
                .build());

    client.initialize(
        SampleUser.CONFIGURED.context(),
        () -> log.add("initialize finished, ready=" + client.isReady()));
  }

  ConfigDirectorClient client() {
    return client;
  }

  SampleLog log() {
    return log;
  }
}

File: samples/configdirector-android/compose/src/main/kotlin/com/configdirector/sample/compose/SampleApplication.kt
package com.configdirector.sample.compose

import android.app.Application
import com.configdirector.AndroidLogger
import com.configdirector.ClientOptions
import com.configdirector.ConfigDirectorClient
import com.configdirector.LogLevel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch

/**
 * The client belongs to the application: one instance, created once, initialized during startup and
 * shared by every screen.
 *
 * Nothing closes it. It lives as long as the process does, and Android reclaims everything when the
 * process ends -- `onTerminate` only runs on an emulator. Call `close()` when an app wants the
 * client gone before that, on sign-out for instance.
 */
class SampleApplication : Application() {

    lateinit var client: ConfigDirectorClient
        private set

    val hasSdkKey: Boolean = BuildConfig.CLIENT_SDK_KEY.isNotEmpty()

    private val scope: CoroutineScope = MainScope()

    override fun onCreate() {
        super.onCreate()

        client = ConfigDirectorClient(
            androidContext = this,
            clientSdkKey = BuildConfig.CLIENT_SDK_KEY.ifEmpty { PLACEHOLDER_SDK_KEY },
            // The SDK logs at WARN by default; turned up here so the connection can be followed
            // in logcat.
            options = ClientOptions.build { logger(AndroidLogger(LogLevel.DEBUG)) },
        )

        scope.launch { client.initialize(SampleUser.CONFIGURED.context) }
    }

    private companion object {
        // The client rejects a blank key, so with none configured the sample runs on a stand-in
        // that the server will not recognize, and says so on screen.
        private const val PLACEHOLDER_SDK_KEY = "no-client-sdk-key-configured"
    }
}
